### PR TITLE
[SPARK-29592][SQL] ALTER TABLE (set partition location) should look up catalog/table like v2 commands

### PR DESCRIPTION
**Pull request by imback82**
### What changes were proposed in this pull request?

Update `AlterTableSetLocationStatement` to store `partitionSpec` and make `ALTER TABLE a.b.c PARTITION(...) SET LOCATION 'loc'` fail if `partitionSpec` is set with unsupported message.

### Why are the changes needed?

It's important to make all the commands have the same table resolution behavior, to avoid confusing end-users. e.g.

```
USE my_catalog
DESC t // success and describe the table t from my_catalog
ALTER TABLE t PARTITION(...) SET LOCATION 'loc' // report set location with partition spec is not supported.
```
### Does this PR introduce any user-facing change?

yes. When running ALTER TABLE (set partition location), Spark fails the command if the current catalog is set to a v2 catalog, or the table name specified a v2 catalog.

### How was this patch tested?

New unit tests

### CODE REVIEW

**Comment by imback82** [comment]
cc: @cloud-fan @rdblue @viirya

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 1]
```diff
@@ -168,8 +168,8 @@ statement
         DROP (IF EXISTS)? partitionSpec (',' partitionSpec)* PURGE?    #dropTablePartitions
     | ALTER VIEW tableIdentifier
         DROP (IF EXISTS)? partitionSpec (',' partitionSpec)*           #dropTablePartitions
-    | ALTER TABLE multipartIdentifier SET locationSpec                 #setTableLocation
-    | ALTER TABLE tableIdentifier partitionSpec SET locationSpec       #setPartitionLocation
+    | ALTER TABLE multipartIdentifier
+        (partitionSpec)? SET locationSpec                              #setTableLocation
```
Comment (imback82) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Please let me know if this continues to be parsed separately.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statements.scala [hunk 2]
```diff
@@ -179,6 +179,7 @@ case class AlterTableUnsetPropertiesStatement(
  */
 case class AlterTableSetLocationStatement(
     tableName: Seq[String],
+    partitionSpec: Option[TablePartitionSpec],
```
Comment (imback82) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statements.scala]: I am piggybacking on the existing `AlterTableSetLocationStatement` (by looking at how `AlterTableSetLocationCommand` is doing). Please let me know if this is not what we want and I will create a separate statement for partition.

**Comment by viirya** [comment]
seems a related test failure.

**Comment by cloud-fan** [comment]
LGTM if tests pass

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala [hunk 3]
```diff
@@ -136,14 +136,21 @@ abstract class SQLViewSuite extends QueryTest with SQLTestUtils {
...
-      assertViewNotSupported(s"ALTER TABLE $viewName SET LOCATION '/path/to/your/lovely/heart'")
+      assertAnalysisExceptionThrown(
+        s"ALTER TABLE $viewName SET LOCATION '/path/to/your/lovely/heart'",
+        s"'$viewName' is a view not a table")
+
+      // For the following v2 ALERT TABLE statements, unsupported operations are checked first
+      // before resolving the relations.
+      assertAnalysisExceptionThrown(
+        s"ALTER TABLE $viewName PARTITION (a='4') SET LOCATION '/path/to/home'",
+        "ALTER TABLE SET LOCATION does not support partition for v2 tables")
```
Comment (imback82) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala]: @cloud-fan, If we need to retain the same behavior to throw an exception with `"'$viewName' is a view not a table"` message, we need to pass `partitionSpec` to `AlterTable`, then check if `partitionSpec` is set inside `AlterTableExec`. I don't think it's necessary, but please let me know if that is preferred.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala [hunk 4]
```diff
@@ -177,9 +184,9 @@ abstract class SQLViewSuite extends QueryTest with SQLTestUtils {
     }
   }
 
-  private def assertViewNotSupported(query: String): Unit = {
+  private def assertAnalysisExceptionThrown(query: String, message: String): Unit = {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala]: nit: can be simply `assertAnalysisError`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala]: looks reasonable to me

Comment (imback82) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala]: I assume you meant the current approach looks reasonable, right?

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala]: yea, the unsupported feature should be reported first.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**